crates/rerun/src/clap.rs

```diff
@@ -68,81 +64,82 @@ pub struct RerunArgs {
     bind: String,
 }
 
+/// [`RerunArgs::init`] might have to spawn a bunch of background tasks depending on what arguments
+/// were passed in.
+/// This object makes sure they live long enough and get polled as needed.
+#[doc(hidden)]
+#[derive(Default)]
+pub struct ServeGuard {
+    tokio_rt: Option<tokio::runtime::Runtime>,
+}
+
+impl Drop for ServeGuard {
+    fn drop(&mut self) {
+        if let Some(tokio_rt) = self.tokio_rt.take() {
+            eprintln!("Sleeping indefinitely while serving web viewer... Press ^C when done.");
+            tokio_rt.block_on(async {
+                tokio::time::sleep(std::time::Duration::from_secs(u64::MAX)).await;
+            });
+        }
+    }
+}
+
 impl RerunArgs {
-    /// Set up Rerun, and run the given code with a [`RecordingStream`] object
-    /// that can be used to log data.
-    ///
-    /// Logging will be controlled by the `RERUN` environment variable,
-    /// or the `default_enabled` argument if the environment variable is not set.
+    /// Creates a new [`RecordingStream`] according to the CLI parameters.
     #[track_caller] // track_caller so that we can see if we are being called from an official example.
-    pub fn run(
-        &self,
-        application_id: &str,
-        default_enabled: bool,
-        run: impl FnOnce(RecordingStream) + Send + 'static,
-    ) -> anyhow::Result<()> {
-        // Ensure we have a running tokio runtime.
-        let mut tokio_runtime = None;
-        let tokio_runtime_handle = if let Ok(handle) = tokio::runtime::Handle::try_current() {
-            handle
-        } else {
-            let rt = tokio::runtime::Runtime::new().expect("Failed to create tokio runtime");
-            tokio_runtime.get_or_insert(rt).handle().clone()
-        };
-        let _tokio_runtime_guard = tokio_runtime_handle.enter();
-
-        let (rerun_enabled, store_info, batcher_config) =
-            crate::RecordingStreamBuilder::new(application_id)
-                .default_enabled(default_enabled)
-                .into_args();
-
-        if !rerun_enabled {
-            run(RecordingStream::disabled());
-            return Ok(());
-        }
+    pub fn init(&self, application_id: &str) -> anyhow::Result<(RecordingStream, ServeGuard)> {
```

Comment:
The ergonomics of getting a tuple back where one part is super important and useful and the other an odd token I have to care around (but only when doing serve) are a bit odd. Even worse if I can `init` with `serve` from a scope now, it will not exit that scope iff I do serve (since the ServeGuard gets dropped). That's quite surprising behavior.
Suggestion for possible alternative: Given that the serve guard can't be killed, we might as well put it on a thread_local variable instead?

Reply:
I tried a lot of things, including putting the guard in a thread local, but the maybe-owned tokio runtime just makes things weird no matter what...

We _do_ want the user to be able to drop the guard, it's the only to shutdown all the web stuff that we've spawned in the background :/